org.openjdk.jmc.console.ext.agent/src/main/java/org/openjdk/jmc/console/ext/agent/manager/wizards/PresetEditingWizardEventPage.java

```diff
@@ -67,20 +91,189 @@ public void createControl(Composite parent) {
+		removeButton = createButton(container, LABEL_REMOVE_BUTTON);
+
+		// set button initial states when no item is selected
+		editButton.setEnabled(false);
+		duplicateButton.setEnabled(false);
+		removeButton.setEnabled(false);
+
+		return container;
+	}
+
+	private Button createButton(Composite parent, String text) {
+		Button button = new Button(parent, SWT.NONE);
+		button.setText(text);
+		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, false));
+		return button;
+	}
+
+	private void bindListeners() {
+		addButton.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				openEventEditingWizardFor(new Event());
+				tableViewer.refresh();
+			}
+		});
+
+		editButton.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				openEventEditingWizardFor((IEvent) tableViewer.getStructuredSelection().getFirstElement());
+				tableViewer.refresh();
+			}
+		});
+
+		duplicateButton.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				// TODO: create a copy properly
+				IEvent original = (IEvent) tableViewer.getStructuredSelection().getFirstElement();
+				IEvent duplicate = new Event();
+				duplicate.setId(original.getId() + ".copy");
+				duplicate.setName("Copy of " + original.getName());
+				preset.addEvent(duplicate);
+				tableViewer.refresh();
+			}
+		});
+
+		removeButton.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				preset.removeEvent((IEvent) tableViewer.getStructuredSelection().getFirstElement());
+				tableViewer.refresh();
+			}
+		});
+
+		tableViewer
+				.addSelectionChangedListener(selectionChangedEvent -> List.of(editButton, duplicateButton, removeButton)
+						.forEach(button -> button.setEnabled(!tableViewer.getStructuredSelection().isEmpty())));
+	}
+
+	private void populateUi() {
+		tableViewer.setInput(preset);
 	}
 
-	// TODO: call this function when "New" or "Edit" button clicked
 	private void openEventEditingWizardFor(IEvent event) {
-		if (DialogToolkit.openWizardWithHelp(new EventEditingWizard(event))) {
-			// TODO: save the modified event to the preset
+		while (DialogToolkit.openWizardWithHelp(new EventEditingWizard(event))) {
+			try {
+				// TODO: save the modified event to the preset
+				preset.addEvent(event);
+			} catch (IllegalArgumentException e) {
+				if (DialogToolkit.openConfirmOnUiThread(MESSAGE_UNABLE_TO_SAVE_THE_PRESET, e.getMessage())) {
+					continue;
+				}
+			}
+
+			break;
 		}
 	}
+
+	private static class EventTableContentProvider extends AbstractStructuredContentProvider
+			implements IContentProvider {
+
+		@Override
+		public Object[] getElements(Object inputElement) {
+			if (!(inputElement instanceof IPreset)) {
+				throw new IllegalArgumentException("input element must be a IPreset"); // $NON-NLS-1$
```

Comment:
Create constant string error message

Reply:
I'd argue this is error message is for developers only as exception like this should never occur to end-user if code is written properly.

Reply:
I was under the impression that we were declaring all constant strings as static class variables, no matter if the message is for the user or developer.

Reply:
Ahh, I don't think this is necessary, nor I've seen people doing this in JMC code base.